### ISSUE TITLE
Add operator.getitemtuple & operator.itemtuplegetter

### ISSUE THREAD

**Issue by dg-pb** [python/cpython#128000]
### Proposal:

```python
from itertools import repeat
from operator import getitemtuple, itemtuplegetter

obj = ['a', 'b']
indices = [0, 1, 2, 3]
defaults = repeat(None)

itg = itemtuplegetter(indices, defaults=defaults)
print(itg(obj))
# ('a', 'b', None, None)

print(getitemtuple(obj, indices, defaults))
# ('a', 'b', None, None)
```

Other form was considered: `itemitergetter()() -> Iterator`.
But its performance has turned out to be non-competitive (see discourse discusssion).
Thus, it is best to follow `tuple` return same as existing `itemtgetter`.

### Has this already been discussed elsewhere?

No response given

### Links to previous discussion of this feature:

https://discuss.python.org/t/itemgetter-split-into-2-objects/69020/33

**Comment by picnixz** [comment]
Ignore this if it has this been already decided on DPO (there are too many messages for a fast overview).

---

Bikeshedding here but can we align the default with whatever `next` does? namely a positional-only argument passed to `itg(object, default)` instead of passing the default at construction time? Or allow both maybe (and reject if defaults have already been specified). My rationale is that you might want to only construct `itemtuplegetter` and, when you want to use defaults, you would then do `itg(x, d)` insteada of `itg(x)`. 

But otherwise, I'm fine with specifying the defaults at construction time, but without a keyword argument (for instance, "defaults" could mean the default values to return and not the filling values). So, we could also use `fill=...` or `fillevalues=...` (we have `fillvalue` for `zip_longest`).

**Comment by dg-pb** [comment]
> Bikeshedding here but can we align the default with whatever `next` does?

The case when one wants to do `itg(obj, defaults)` is covered by `getitemtuple(obj, items, defaults)`.
`itg(obj, defaults)` would have no advantages over the former as it would need to do large part of the input parsing that is done in `__init__` at `__call__-time`.

For the convenience of binding, one can use `partial(getitemtuple, Placeholder, items)`. Would be marginally slower than specialisation in `itemtuplegetter`, but I think it is a happy middle.

> But otherwise, I'm fine with specifying the defaults at construction time, but without a keyword argument (for instance, "defaults" could mean the default values to return and not the filling values). So, we could also use `fill=...` or `fillevalues=...` (we have `fillvalue` for `zip_longest`).

Could you clarify? `defaults` are "default values to return" if item not found in object.
I.e.
```python
items = ['c', 'b']
defaults = [0]

getitemtuple(obj, items, defaults)
# Equiv
result = list()

for i in range(len(items)):
    idx = items[i]
    if i < len(defaults):
        try:
            result.append(obj[idx])
        except (KeyError, IndexError)
            result.append(defaults[i])
    else:
        result.append(d[idx])
```

**Comment by dg-pb** [comment]
In this initial version, there is a restriction on the fact that defaults can only be provided for "all" or "leading" items.
And there is no way to have a default for 2nd, but not the first item.

To overcome this, in many cases one can adjust the order of `items`.

While this is a good solution that can address many cases, there is an option for `NoDefault` sentinel extension in the future, which would allow having both "desired order" and "sparse defaults".

However, I think time will tell whether there is a need for such and this initial concept already covers many cases.

**Comment by picnixz** [comment]
If you want defaults to skip something, use a dict with keys being the indices (I think this should work):

```py
if isinstance(defaults, Mapping):
	for i in range(len(items)):
	    idx = items[i]
	    try:
	        result.append(obj[idx])
	    except LookupError:
	        result.append(defaults[i])
```

Sparse defaults should assume that you don't expect other indices to fail. Non-sparse defaults would use your above example.

**Comment by dg-pb** [comment]
Thought about this and unfortunately it imposes undesirable restriction - `dict` only supports hashable keys.

Also, `dict` access is very slow compared to sequence and while it is nothing major, but this specific component seems to be largely liked due to its performance, thus this might be a dealbreaker for many.

Of course, as you suggested one can have an option to use `dict` or not, but this just leads to further restriction and still leaves user with needing to pick some of the features without the ability to have "all".

Thus, this leads to more complexity and specialisation.
While, `NoDefault` placeholder is not as elegant as `dict`, but it would provide a definite "have it all" solution.

Personally, I like `sentinel` solutions. I think part of the discomfort with these can be addressed to lack of implementation standard & convenient toolkit.

I suspect that this might change to some degree given there was a place in `CPython`, where one can just add 1-line to create new `singleton-sentinel`.

**Comment by picnixz** [comment]
Oh right. I tend to forget that `itemgetter` is not restricted to hashable keys but can be done for any type of keys. Forget the dict suggestion.

**Comment by rhettinger** [comment]
Sorry, I don't support this being added to the operator module. It is well beyond the module's intended scope, but its biggest issue is that it creates confusion from too many similar ways to do it. Tell someone that there is an itemgetter, getitemtuple, and itemtuplegetter and see if they can deduce in the intended distinctions and know which to choose.

Also, the performance aspect of the newsgroup discussion is a red herring. It is reasonably possible that  the JIT efforts will make plain Python code nearly as fast as what this would offer.  There is no longer sufficient motivation to invent more exotic ways to avoid a simple `lambda` or `def`.

Recall that the reason Guido allowed `itemgetter` and `attrgetter` was to support writing nice looking key-functions.  It was never intended to evolve beyond that.

**Comment by picnixz** [comment]
> Tell someone that there is an itemgetter, getitemtuple, and itemtuplegetter and see if they can deduce in the intended distinctions and know which to choose.

(I personally couldn't at first; I don't use a lot of `itemgetter` or `attrgetter` personally, so I'm also +0.)

**Comment by dg-pb** [comment]
Gauging sentiment in discussions I didn't think there is going to be a lot of resistance.

All good points above, I hope I can convey a bit more complete rationale for this by answering them.

---

> its biggest issue is that it creates confusion from too many similar ways to do it. Tell someone that there is an itemgetter, getitemtuple, and itemtuplegetter and see if they can deduce in the intended distinctions and know which to choose.

I don't think it is that confusing.
It isn't 3 objects.
It is 2 + 2:
1. `getitem` (and its `itemgetter`)
2. `getitemtuple` (and its `itemtuplegetter`)

The mental model is that `itemgetter` should be used for 1 item and `itemtuplegetter` for many items.

`itemgetter` should never have allowed multiple items.
I don't think unpredictability in its return value was a good decision.

`itemgetter` does 2 things, but doesn't do the second one well. At the same it has blocked the way for other reliable tool addition for multi-item cases such as this.

It currently is prone to bugs when constructing it with an unknown number of items without a good alternative solution: https://github.com/python/cpython/pull/124788/files#diff-8657d1a12c81ea9da8c75d3de3827fc6f1e033339453e9587f1ec2f5dfc5111dR314

---

Another motivation for this was also to cover the `multi-item` part of: https://discuss.python.org/t/indexable-get-method-1-2-3-get-4-none/53430/3
So https://discuss.python.org/t/allowing-missing-item-or-attributes-in-operators-itemgetter-and-attrgetter/66256/4 would cover the `single-item` case and this would cover the `multi-item`.

---

> It is well beyond the module's intended scope

I think it could be fairly reasonable to include "abstract object layer" as a possible secondary scope.

There are some functionality/features desirable by community and there are occasional requests to extend various collections with new methods.

Naturally, it is extremely unlikely that something will be worth the effort to add a new method to say `list`.

However, adding small utilities to `operator` when there is a good case would seem reasonable to me. Such do not have any impact on object protocols and can often cover more than 1 object. E.g. `list.get` only applies to `list` while `getitem(..., default=)` would apply to most of the objects that support `__getitem__`.

Having said that, I don't think this should be abused and only allow for additions that cover a sufficiently large scope of needs with minimal additions.

> Also, the performance aspect of the newsgroup discussion is a red herring. It is reasonably possible that the JIT efforts will make plain Python code nearly as fast as what this would offer.

Although I see that theoretically it could come close, I doubt it will ever match.
Performance is only a fraction of the rationale for this, but knowing that it will always remain the most performant solution, even if by a small percentage, is a positive.

Some benchmarks:
```python
lst = [None] * N
idxs = list(range(N))
noidxs = [N] * N
# N = 2 (ns)
%timeit getitemtuple(lst, idxs)           # C: 190, Py: 370
%timeit getitemtuple(lst, noidxs, idxs)   # C: 550, Py: 1000
# N = 100K (ms)
%timeit getitemtuple(lst, idxs)           # C: 1.7, Py: 2.9
%timeit getitemtuple(lst, noidxs, idxs)   # C: 18,  Py: 31
```

So currently it is 40-50% faster.
So even if, say, JIT makes Python version only 10% slower, this would still be a 10% faster solution that can be used to simplify code in variety of cases.

E.g. This would be a good option for partial defaults.
In this case, it solves a fairly common problem, where otherwise there is no convenient solution.
```
sys_argv = [0, 1]

b, a = getitemtuple(sys_argv, [2, 1], ['b_dflt'])
(a, b)    # 0, 'b_dflt'

getitemtuple(sys_argv, [], ['b_dflt'])
# Error, first argument does not have a default
```

---

To sum up, I think this proposal is quite well balanced addition in several aspects:
a) Straightens out design flaws of `itemgetter`
b) Covers `list.get` feature requests that came up in the past and I predict will come up over and over again from time to time.
c) Provides convenience by abstracting "multiple-item-case" `try-except-KeyError/IndexError-default` for a wider object scope, which is a fairly common and quite verbose pattern. And this is useful in a variety of different cases - both mappings and sequences.
d) Although I can see that JIT might come close, I doubt that it would ever fully match the performance of this addition.

**Comment by dg-pb** [comment]
> Recall that the reason Guido allowed `itemgetter` and `attrgetter` was to support writing nice looking key-functions. It was never intended to evolve beyond that.

I appreciate the absence of initial intent, but I think this has already been affected by https://www.hyrumslaw.com to the extent that it might make sense to improve the situation by solving some unintended consequences.

**Comment by picnixz** [comment]
I'm still neutral on this one, but I'd like to reply on this one:

> itemgetter does 2 things, but doesn't do the second one well

Err... maybe I'm biased, but `itemgetter` is for me just a projector onto one or more components, with an implicit flattening when you only have one component. Granted that we could have named it `itemSgetter` to indicate a plural form and to have two distinct objects, but a unified interface could be fine for that. More precisely, `itemgetter` is (AFAIK and AFAICT) equivalent to a projector 

$$
	\pi_{i_1,\ldots,i_m}\colon \prod_{i=1}^nX_i\longrightarrow \prod_{j=1}^m X_{i_j},\quad (x_1,\ldots,x_n)\mapsto (x_{i_1},\ldots,x_{i_m}).
$$

Now, when $m = 1$, the projector is a simple component projection, and by convention, it would return a scalar (i.e. an element of $X_{i_1}$).

**Comment by dg-pb** [comment]
> `itemgetter` is for me just a projector onto one or more components
> Now, when m=1, the projector is a simple component projection, and by convention, it would return a scalar

Currently `itemgetter` is exactly that.

And part of this is to address impracticalities that arise from "one or more" behaviour.

I don't think there are any issues with theory here, that point is about practical implications.

Current practical issue is exactly the uncertainty of `return` given it isn't known whether m==1 or not.

And this starts to cause further issues with extension attempts, such as addition of `default`. In short, things start becoming much more complex than they could be otherwise.

---

2 implementations - one for scalar and one for vector - would simplify things in many practical aspects.

However, as `itemgetter` is already a hybrid, the possible next best thing is to keep it for m==1 and introduce `iterable-version`.

**Comment by picnixz** [comment]
> I don't think there are any issues with theory here, that point is about practical implications.

Ah, I misunderstood your "doesn't do the second one well" thinking that you thought there were flaws in the multi-dimensional case apart from the return type.

> Current practical issue is exactly the uncertainty of return given it isn't known whether m==1 or not.

Wouldn't, in most of the cases, be the caller responsibility? Alternatively, one can add a boolean on those objects to remember whether it was `m = 1` or not (so to distinguish whether `(x, y, z)` is actually a projection of `((x, y, z),)` (1 three-dimensional component) or if it's a projection onto 3 one-dimensional components). Granted this could looks like a hack though but wouldn't it be sufficient for most of the cases? (I don't remember if the `itemgetter` object doesn't already remember the indices).

> And this starts to cause further issues with extension attempts, such as addition of default. In short, things start becoming much more complex than they could be otherwise.

Whether defaults should be supported or not is something I don't know. I would say "no" because I actually don't expect `__getitem__` to actually have a default value. But that's perhaps something separate. 

I however see value in implementing this now if we were to have a nullish coalescing (I think it's PEP-505). If PEP-505 is accepted, we can think of `__getitem__` as supporting possible "default" values, so *semnatically* it makes sense to have an interface reflecting this in `operator`, but for now defaults do not *semantically* make sense IMO (again this is because of the fact that it's something in the `operator` module which essentially either relate to primitive operations or dunders). OTOH, it only becomes more complex *if* we use this interface and/or decide to add new features. 

An alternative would be to suggest this to `more-itertools` and see if there are use cases? While your partialmethod use-case is valid, it's only one amongst few.

---

I'm still staying on my -0.

**Comment by dg-pb** [comment]
> I however see value in implementing this now if we were to have a nullish coalescing

Agreed, this would complement PEP505, but I don't think that non-existence of PEP505 has a negating effect on this.

> While your partialmethod use-case is valid, it's only one amongst few.

There were few threads related to this and people seem to agree that this nuance is causing a certain discomfort:
1. https://discuss.python.org/t/itemgetter-always-tuple-keyword-arg/67198
2. https://discuss.python.org/t/itemgetter-split-into-2-objects/69020
3. https://discuss.python.org/t/allowing-missing-item-or-attributes-in-operators-itemgetter-and-attrgetter/66256/9

And I think there is a general agreement that input-dependent-return-structure is not a very good practice.

The way I see it, current situation of this hardly passes design quality standard for standard library.

**Comment by erlend-aasland** [comment]
I'm aligned with Raymond's objection. Count me as -1.

**Comment by ncoghlan** [comment]
I encouraged this proposal specifically because of the type ambiguity in the output of the existing functions.

This isn't a new feature, it's a more explicit spelling of the existing tuple support in `itemgetter`. Unlike the existing spelling, it doesn't implicitly change return type from `tuple[Any]` to `Any` when you go from retrieving two values to retrieving one value, so there's no ambiguity between the function returning multiple values and it returning a single value that happens to be a tuple.

**Comment by ncoghlan** [comment]
The alternative would be to document that `itemgetter` should never be used with `*args` without ensuring that `args` contains more than one item.

It seemed harsh to do that to users when we could instead provide an alternative better suited to that case.

**Comment by rhettinger** [comment]
> The alternative would be to document that itemgetter should never be used with *args without ensuring that args contains more than one item.

To me, that is a reasonable outcome.  We've had the current API for two decades and no one has wanted anything different until now.  It is possible that the only two people in the world who care about this corner case are already in this thread ;-).   Joking aside, this isn't unusual, see `min()` and `max()` for example.  And if the need does arise, it isn't hard to test for the case where `len(args)==1`.

Sorry to be a strong -1 on this, but I think it would be end-up cluttering the module and be a permanent wart that results in confusing users more than it helps them. We all have had experience with modules that have too many variants of the same function, each with subtle differences between them, and where the distinctions only rarely matter.

**Comment by dg-pb** [comment]
> Sorry to be a strong -1 on this, but I think it would be end-up cluttering the module and be a permanent wart that results in confusing users more than it helps them.

Ok fair. Then what would you say about introducing keyword argument instead, which by default is `False`, but if `True`, then returns `tuple` even if only 1 item was provided?

**Comment by dg-pb** [comment]
> Joking aside, this isn't unusual, see `min()` and `max()` for example.

For `min()` and `max()` return structure is the same for both variants, while `itemgetter` case changes structure and currently the user doesn't have sufficient control.

However, it is possible to align `itemgetter` with `min` max by introducing a keyword. I.e.:
```python
min(1, 2, 3) -> 1
min([1, 2, 3]) -> 1

itemtuplegetter(obj, 1) -> Scalar
itemtuplegetter(obj, 1, 2, 3) -> tuple

itemtuplegetter(obj, [1], itermode=True) -> tuple
itemtuplegetter(obj, [1, 2, 3], itermode=True) -> tuple
```

**Comment by ncoghlan** [comment]
Considering the "documented design pattern" case over the "API change" case:

```python
def itemtuplegetter(items: Iterable[Any]):
    seq_items = tuple(items)
    if not seq_items:
        # Alternative: `return lambda _: ()` if requesting zero items is permitted
        raise ValueError("Must request at least one item")
    if len(seq_items) == 1:
        item = seq_items[0]
        return lambda obj: (obj[item],)
    return itemgetter(*seq_items)
```

The fact there are two potentially valid behaviours for the "zero items requested" scenario is  pushing me back towards @rhettinger's "not every three-line function needs to be a builtin" perspective.

With a documented recipe, folks who have this need can tailor their wrapper function to their specific case (including deciding if and how they want to handle default values), which simplifies the API for that particular application, whereas trying to accommodate all the potential variations in a stdlib API makes that API *harder* to use instead of easier (see also: "Why doesn't `itemgetter` support default values?" and "Why does the `more-itertools` PyPI package exist?")

So apologies from me, too @dg-pb - I genuinely thought this was a justifiable addition, but I'm afraid I hadn't adequately considered what the "provide better documentation for the 'variable number of lookup items' case " alternative would look like (and it's not as bad as I thought it would be, which changes my opinion on the viability of that approach).

I guess my comment from the dpo thread also still applies: it's likely worth checking with the `more-itertools` maintainers to see if they think this would be a worthwhile addition to their set of stdlib-adjacent utility APIs.

**Comment by dg-pb** [comment]
> Considering the "documented design pattern" case over the "API change" case:

Issue with this is that it becomes unserialisable.
I used it for a while, but removed it not long after.

and `more_itertools` is going to be the same speed as doing `list(map(getitem, repeat(obj), items))` or something similar.

Personally, I don't see any good alternatives.

I don't think there is an issue with coming up with solution, the issue is that those solutions are not satisfactory in the light of how it could be if `itemgetter` was improved a bit.

Those who encounter this will just adapt base on use case as they did up to this point: maybe current `itemgetter` will do, maybe recipe above with it, maybe comprehension, maybe loop statement if defaults are needed.

> So apologies from me

No worries, one can never know how things will turn out. :)

**Comment by erlend-aasland** [comment]
Rebranded as a docs issue, per Raymond's and Alyssa's comments.

**Comment by dg-pb** [comment]
Just to make sure whether this is a no-go as well before I make my peace with this.

Would a flag be unfavourable option as well?
```
# The way it is now (itermode=False)
ig = itemgetter('a', 'b', 'c')

# The way it is now (itermode=False) + default
# Single default for all items
ig = itemgetter('a', 'b', 'c', default=None)

# itermode=True (only 1 positional-iterable allowed)
ig = itemgetter(('a', 'b', 'c'), itermode=True)

# itermode=True (default is also an iterable)
ig = itemgetter(('a', 'b', 'c'), default=(None,), itermode=True)
```

@ncoghlan, @rhettinger, would this extension also be undesirable?

**Comment by picnixz** [comment]
> Would a flag be unfavourable option as well?

AFAIK, we usually try to avoid parametrized return types (it's already bad that we have a different return type depending on the inputs). One reason is that it makes typesheds definitions more complex. I'm also not aware of this practice elsewhere in the library.

**Comment by dg-pb** [comment]
> it's already bad that we have a different return type depending on the inputs

Given the current situation I don't see how this makes things any more complex:

Before - return type is dependent on inputs
After - return type is dependent on inputs

**Comment by picnixz** [comment]
After: it's *doubly* dependent on inputs I'd say and it introduces another parameter that newcomers need to learn. So it *does* makes things more complex. We also need to update the docs and this would make things more confusing IMO. It mixes too many things in one interface that wasn't meant to be that complex.

**Comment by dg-pb** [comment]
Well obviously I was referring to what you have said, and not the fact that additional feature does not introduce additional complexity.

It is all obvious what you have said.

I would much more appreciate balanced view which takes into account costs versus benefits. Now you are just fighting "against" instead of "what could be best".

**Comment by picnixz** [comment]
I'm not fighting against. I'm explaining why this is likely to be rejected and why I see how this is more complex.

Having a return type depends on one input is sufficiently complex. Adding another layer of complexity on top of that is not the best IMO. I'm also considering the costs (docs, typing, teaching) vs the benefits. This can be transformed into a one-liner if users want to have a parametrized return-type but I don't think merging an interface is a solution (if I were to make a choice between a keyword and a new function, I would prefer new functions instead, at least this decouples the abstraction; but if adding new functions is a no go, then I would prefer rejecting the feature altogether).

**Comment by erlend-aasland** [comment]
@dg-pb, this is not productive; why do you sustain your pushy behaviour despite Alyssa's and Raymond's suggested documentation updates?

**Comment by dg-pb** [comment]
> Adding another layer of complexity on top of that is not the best IMO

Well yes, this does add an extra layer of complexity.

If complexity of return depending on 1 input is 100%, complexity of return type depending on 2 inputs is maybe 120%. This is already there, so this does not add up and complexity increase is marginal.

> This can be transformed into a one-liner if users want to have a parametrized return-type

There are issues with one-liner. Proper implementation would be more robust and performant.

Also, this is not only about return type, but also incorporating defaults. There is another thread of adding `default` to `itemgetter` so this attempts to cover that too.

> (docs, typing, teaching)

docs - yes. typing + 20%, teaching - a bit.

> @dg-pb, this is not productive; why do you sustain your pushy behaviour despite Alyssa's and Raymond's suggested documentation updates?

That has nothing to do with documentation updates.
Am I not allowed to ask for opinion regarding alternative solution?

I don't think I am being pushy.
I asked a question fairly gently and was accepting to any kind of solid reasoning.

Now I am just responding back. And the quality of my responses largely reflect the quality of stuff that was thrown at me.

**Comment by rhettinger** [comment]
On behalf of the reviewers, I'm marking this feature request as declined.  We don't see it as meeting the high bar for a standard library addition. If you want to write your own variants in pure Python, it is not difficult to do.  Thanks for the suggestion, but we will pass on it.
